File: elixir/pascals-triangle/pascals_triangle.exs
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num), do: for n <- 0..num - 1, do: for k <- 0..n, do: rows(n,k)

  def rows(_, 0), do: 1
  def rows(n, k), do: n |> rows(k-1)  |> Kernel.*(n + 1 - k) |> Kernel.div(k)

end

File: julia/pascals-triangle/pascals-triangle.jl
using Match

function triangle(num::Int)
    function cell(n, k) 
        @match (n, k) begin
            (_, 0) => 1
            (n, k) => div(cell(n, k - 1) * (n + 1 - k), k)
        end
    end
    num < 0 ? throw(DomainError(nothing)) : 
      [[cell(row, col) for col in  0:row] for row in 0:(num - 1)]
end
